CRM/Upgrade/Incremental/php/TimezoneRevertTrait.php

```diff
@@ -0,0 +1,302 @@
+      MODIFY COLUMN `registration_start_date` datetime DEFAULT NULL COMMENT 'Date and time that online registration starts.',
+      MODIFY COLUMN `registration_end_date` datetime DEFAULT NULL COMMENT 'Date and time that online registration ends.';";
+
+    $upgrade = new CRM_Upgrade_Form();
+    $upgrade->source($sql, TRUE);
+
+    return TRUE;
+  }
+
+  /**
+   * If a user edited an `Event` in the UI while running 5.47.alpha1 - 5.47.2, then `revertEventDates`
+   * won't be good enough. In particular:
+   *
+   * - It's likely to have activated the DST bug (based on the current-user's TZ).
+   * - The user could have filled-in/corrected the values of `event_tz` and/or each `TIMESTAMP` column.
+   *
+   * The algorithm reads backup values (eg `start_date_ts_bak`) and rewrites live values (eg `start_date`)
+   *
+   * It uses a heuristic approach to cleaning DST error ("skew"). This requires a _representative_
+   * timezone ("skewTz"). It should not be necessary to know the exact TZ of every edit -- as long as all TZs
+   * have similar DST rules. For example:
+   *
+   *   - Most locales in US+CA have the same DST rules. (`America/Los_Angeles` and `America/Chicago` are equally representative.)
+   *   - Most locales in Europe have the same DST rules. (`Europe/Helsinki` and `Europe/Berlin` are equally representative.)
+   *   - Most locales in Australia have the same DST rules.
+   *
+   * By default, this will borrow the current user (sysadmin)'s timezone as the representative skewTz.
+   * This can be overridden with env-var `CIVICRM_TZ_SKEW`.
+   *
+   * @param \CRM_Queue_TaskContext $ctx
+   * @return bool
+   */
+  public static function convertModifiedEvents(CRM_Queue_TaskContext $ctx = NULL): bool {
+    $skewTz = self::pickSkewTz();
+    $mysqlTz = '+0:00';
+    $restoreMysqlTz = static::swapTz($mysqlTz);
+
+    $columns = [
+      'start_date' => 'start_date_ts_bak',
+      'end_date' => 'end_date_ts_bak',
+      'registration_start_date' => 'registration_start_date_ts_bak',
+      'registration_end_date' => 'registration_end_date_ts_bak',
+    ];
+
+    [$lowLogId, $highLogId] = self::findLogRange(
+      '5.47.alpha1',
+      (static::class === CRM_Upgrade_Incremental_php_FiveFortySeven::class) ? '5.47.3' : '5.48.beta2'
+    );
+
+    $eventModTimes = CRM_Utils_SQL_Select::from('civicrm_log')
+      ->select('entity_id, max(modified_date) as modified_date')
+      ->where('entity_table = "civicrm_event"')
+      ->where('id >= #lowLogId AND id <= #highLogId', ['lowLogId' => $lowLogId, 'highLogId' => $highLogId])
+      ->groupBy('entity_table, entity_id')
+      ->execute()
+      ->fetchMap('entity_id', 'modified_date');
+    $events = CRM_Utils_SQL_Select::from('civicrm_event')
+      ->select(['id', 'event_tz_bak'])
+      ->select(array_values($columns))
+      ->where('id in (#IDS)', ['IDS' => array_keys($eventModTimes)])
```

Comment:
I get a syntax error here during one of my tests, I presume because I didn't edit any events after upgrading to 5.47.2.

`SELECT id, event_tz_bak, start_date_ts_bak, end_date_ts_bak, registration_start_date_ts_bak, registration_end_date_ts_bak FROM civicrm_event WHERE (id in ()) [nativecode=1064 ** You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '))' at line 3]`

Reply:
Yup, I can reproduce that. Fixed.